docs/recipes/immutability-helpers.md

```diff
@@ -0,0 +1,274 @@
+# Immutability Helpers
+
+Redux is a tiny pattern that represents states as immutable objects. Redux was originally designed for React. Most Redux concepts, such as pure functions, are centered around the React ecosystem. Nowadays Redux is not directly related to React.
+
+The cornerstone of Redux is immutability. Immutabilty is an amazing pattern to minimise unpredictable behaviour in our code. We're not going to cover functional programming in this article. However we're going to look at very useful packages that are called "immutability helpers".
+
+## The Problem
+
+Most developers have to deal with, so called, "deep objects" and most important follow the immutability concept, when it comes to changing the value of some deeply nested property. Given the following code:
+
+```ts
+export interface Task {
+  title: string;
+  dates: {
+    startDate: string;
+    dueDate: string;
+  };
+}
+
+export interface TrelloStateModel {
+  tasks: {
+    [taskId: string]: Task;
+  };
+}
+
+@State<TrelloStateModel>({
+  name: 'trello',
+  defaults: {
+    tasks: {}
+  }
+})
+export class TrelloState {}
+```
+
+Let's imagine that we're faced with the task of changing the `dueDate` property:
+
+```ts
+export class UpdateDueDate {
+  static readonly type = '[Trello] Update due date';
+  constructor(public taskId: string, public dueDate: string) {}
+}
+```
+
+Let's see how we would implement the `updateDueDate` action handler:
+
+```ts
+export class TrelloState {
+
+  @Action(UpdateDueDate)
+  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
+    ctx.setState(state => ({
+      tasks: {
+        ...state.tasks,
+        [action.taskId]: {
+          ...state.tasks[action.taskId],
+          dates: {
+            ...state.tasks[action.taskId].dates,
+            dueDate: action.dueDate
+          }
+        }
+      }
+    }));
+  }
+
+}
+```
+
+Although this code will work but unfortunately it's complicated to maintain and accompany. It's not self-descriptive and will be daunting for upcoming developers.
+
+## Solutions
+
+There are different ways to stay away from such code. This can be accomplished with the use of different packages that may be suitable for your needs.
+
+### State Operators
+
+[State operators](../advanced/operators.md) are the first immutability helpers that NGXS provides out of the box. The `patch` operator will become your best friend in case of choosing state operators as your immutability helpers. Let's see how we could re-write the above code by dint of the `patch` operator:
```

Comment:
```suggestion
[State operators](../advanced/operators.md) are first-class immutability helpers that NGXS provides out of the box. The `patch` operator will become your best friend in case of choosing state operators as your immutability helpers. Let's see how we could re-write the above code with the help of the `patch` state operator:
```